### PR TITLE
Buy from Project Pool

### PR DESCRIPTION
**Pull request by ladytrekker**
## Description

This PR integrates the redeem flow into Carbonmark (Buy from Project Pool)

Still, the pages are not accessible for the users.
See the follow-up ticket https://github.com/KlimaDAO/klimadao/issues/1219

### FEEs
This PR integrates the changes of this PR here too: https://github.com/KlimaDAO/klimadao/pull/1204
=> if a project is the default of a pool => there are less fees
=> get the final costs from RA contract with either `getSourceAmountDefaultRedeem` or `getSourceAmountSpecificRedeem` 
=> see the fee breakdown in corresponding ticket of this PR !  https://github.com/KlimaDAO/klimadao/issues/1127

### Price
There is an issue with the correct price data of a **default** project pool from the Carbonmark API.
Assumption is that the API always gives back the price for a specific project, even for default projects.
Please read the discussion here:
https://github.com/KlimaDAO/klimadao/issues/1156

### PREVIEW LINKS
- BCT  (**default project**) https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-191-2008/purchase/pools/bct
- UBO  (**default project**) https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-1140-2015/purchase/pools/ubo
- NBO (**default project**) https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-981-2014/purchase/pools/nbo
- NCT https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-981-2014/purchase/pools/nct

**Equivalent to the App:**

Note: In the dApp the difference between a default or specific redeem is not implemented.
The dApp always calculates the final costs with `getSourceAmountSpecificRedeem` also for default projects.

- BCT https://app.klimadao.finance/#/redeem?pool=bct&projectTokenAddress=0xb139c4cc9d20a3618e9a2268d73eff18c496b991
- UBO https://app.klimadao.finance/#/redeem?pool=ubo&projectTokenAddress=0xd6ed6fae5b6535cae8d92f40f5ff653db807a4ea
- NBO https://app.klimadao.finance/#/redeem?pool=nbo&projectTokenAddress=0xabfd6760151c7f9361cc5b03bbeebe2d7c0251da
- NCT https://app.klimadao.finance/#/redeem?pool=nct&projectTokenAddress=0xabfd6760151c7f9361cc5b03bbeebe2d7c0251da

### Design Changes
- I added Asset Details with the token the user received (not in the designs)
- I added a Success Modal (missing in the designs, also not possible to render a complete page)

### OPEN QUESTIONs
- [x] **QUESTION:** Also here double the swap fee for BCT (same as for Retire?)? => yes
- [x] Waiting for the integration of https://github.com/KlimaDAO/klimadao/issues/1160
- [x] For the costs calculation => also here differentiate between default and specific costs

## Related Ticket

Resolves https://github.com/KlimaDAO/klimadao/issues/1127

## Changes

| Before  | After  |
|---------|--------|
| Bildschirmfoto 2023-06-08 um 15 57 32 | Bildschirmfoto 2023-06-08 um 15 55 55 |

## Checklist

- [x] I have run `npm run build-all` without errors
- [x] I have formatted JS and TS files with `npm run format-all`

### CODE REVIEW

File: carbonmark/components/pages/Project/Purchase/Pool/AssetDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      <div className={styles.totalsText}>
+        <Text color="lightest">{t`Available to purchase`}</Text>
+        <Text t="h5">
+          {availableTonnes} {t`Tonnes`}
+        </Text>
+      </div>
+      <div className={styles.totalsText}>
+        <Anchor
+          className={styles.iconAndText}
+          href={`https://polygonscan.com/address/${props.price.poolTokenAddress}`}
```
Comment (Atmosfearful) [inline carbonmark/components/pages/Project/Purchase/Pool/AssetDetails.tsx:55]: The `token` view that polygonscan offers is a little more useful for these
```suggestion
          href={`https://polygonscan.com/token/${props.price.poolTokenAddress}`}
```

**Review by Atmosfearful — APPROVED** [review]
looks goooooood

**Review by jabby09 — CHANGES_REQUESTED** [review]
@ladytrekker it looks to me like there is a misc character in the Network fees summary line?

Screenshot 2023-06-14 at 4 33 34 PM

**Review by jabby09 — COMMENTED** [review]
@ladytrekker...this link https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-981-2014/purchase/pools/nct

I believe this IS NOT the default NCT project, however, we are showing fee for KlimaDAO contracts (the RA). I thought for 'BUY' from a pool listing, there are no RA fees?

**Review by jabby09 — CHANGES_REQUESTED** [review]
@ladytrekker on this link https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-191-2008/purchase/pools/bct

Price is $0.91, I entered 1 tonne, my total cost is 0.687??
Screenshot 2023-06-14 at 4 42 07 PM

## UPDATE

As mentioned above in the description:
There is an issue with the price. The total costs are correct and is what the user confirms and pays.

**Review by jabby09 — COMMENTED** [review]
@ladytrekker When I select a purchase amount (tonnes) and then click CONTINUE I seem to be skipping the APPROVE step in the Confirm Purchase modal, and if I click SUBMIT on that form I get an Error

## UPDATE

Not sure about this.
Please check your allowances your wallet has here: https://polygonscan.com/tokenapprovalchecker
And delete them if needed.

**Review by jabby09 — COMMENTED** [review]
@ladytrekker your note on a success modal and this screenshot (I've not been able to complete a transaction yet)
image

you mention there was no design for this, although I see a [receipt design here](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/%2Fprojects?type=design&node-id=11-59688&t=taQ8lJOsmXDfW8pu-4)

plus wouldn't it just match the design of buy from seller listing which we already have?

**Comment by ShimonD-KlimaDAO** [comment]
1. The image at the top is not rounded in the corners and doesn't extend to the edges. Below is actual vs figma:
image
image

2. I'm not able to reconcile the swap fee of 0.3%. How is it calculated exactly? 
The 0.03% double swap fee applies to every token, since every swap has to go through Klima and we don't have co2/usdc pools with deep liquidity. For example, we have the following two amounts but the fee doesn't linearly increase, and I'm getting a different result when calculating myself.
image
image

3. "Processing" modal doesn't match figma
image

Figma: 
image

4. My purchase from pool does not show in the Activity feed in my portfolio page.

5. "Total price" should be "Total Price" to match figma.

6. Total cost for BCT [here](https://carbonmark-git-lady-redeem-klimadao.vercel.app/projects/VCS-191-2008/purchase/pools/bct) doesn't make sense. It's less than the price per tonne.
image

File: carbonmark/lib/actions.redeem.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        getAddress(params.paymentMethod),
+        getAddress(params.pool),
+        utils.parseUnits(
+          params.maxCost,
+          getTokenDecimals(params.paymentMethod)
+        ),
+        utils.parseUnits(params.quantity, 18),
+        0, // correct here ??
+        0 // correct here ??
+      );
```
[the remark this answers is not in the document]
Comment (ladytrekker) [reply carbonmark/lib/actions.redeem.ts:77]: done

File: carbonmark/components/pages/Project/Purchase/Pool/AssetDetails.tsx (hunk 1 above)
Comment (ladytrekker) [reply carbonmark/components/pages/Project/Purchase/Pool/AssetDetails.tsx:55]: done

**Review by jabby09 — APPROVED** [review]

**Pull request merged**